teuthology/task/ansible.py

```diff
@@ -0,0 +1,225 @@
+                raise
+        else:
+            raise TypeError(
+                "playbook value must either be a list, URL or a filename")
+        log.info("Playbook: %s", self.playbook)
+
+    def get_inventory(self):
+        """
+        Determine whether or not we're using an existing inventory file
+        """
+        self.inventory = self.config.get('inventory')
+        etc_ansible_hosts = '/etc/ansible/hosts'
+        if self.inventory:
+            self.inventory = os.path.expanduser(self.inventory)
+        elif os.path.exists(etc_ansible_hosts):
+            self.inventory = etc_ansible_hosts
+        return self.inventory
+
+    def generate_hosts_file(self):
+        """
+        Generate a hosts (inventory) file to use. This should not be called if
+        we're using an existing file.
+        """
+        hosts = self.cluster.remotes.keys()
+        hostnames = [remote.name.split('@')[-1] for remote in hosts]
+        hostnames.sort()
+        hosts_str = '\n'.join(hostnames + [''])
+        hosts_file = NamedTemporaryFile(prefix="teuth_ansible_hosts_",
+                                        delete=False)
+        hosts_file.write(hosts_str)
+        hosts_file.flush()
+        self.generated_inventory = True
+        self.inventory = hosts_file.name
+
+    def generate_playbook(self):
+        """
+        Generate a playbook file to use. This should not be called if we're
+        using an existing file.
+        """
+        for play in self.playbook:
+            play['hosts'] = 'all'
+        pb_buffer = StringIO()
+        pb_buffer.write('---\n')
+        yaml.safe_dump(self.playbook, pb_buffer)
+        pb_buffer.seek(0)
+        playbook_file = NamedTemporaryFile(prefix="teuth_ansible_playbook_",
+                                           delete=False)
+        playbook_file.write(pb_buffer.read())
+        playbook_file.flush()
+        self.playbook_file = playbook_file
+        self.generated_playbook = True
+
+    def begin(self):
+        super(Ansible, self).begin()
+        self.execute_playbook()
+
+    def execute_playbook(self, _logfile=None):
+        """
+        Execute ansible-playbook
+        """
+        environ = os.environ
+        environ['ANSIBLE_SSH_PIPELINING'] = '1'
+        environ['ANSIBLE_ROLES_PATH'] = "%s/roles" % self.repo_path
+        args = self._build_args()
+        command = ' '.join(args)
+        log.debug("Running %s", command)
+
+        out_log = self.log.getChild('out')
+        out, status = pexpect.run(
+            command,
+            logfile=_logfile or LoggerFile(out_log, logging.INFO),
+            withexitstatus=True,
+            timeout=None,
+        )
+        if status != 0:
+            raise CommandFailedError(command, status)
+
+    def _build_args(self):
+        """
+        Assemble the list of args to be executed
+        """
+        fqdns = [r.name.split('@')[-1] for r in self.cluster.remotes.keys()]
```

Comment:
Ah, that's right.  We should probably document that behavior in the docstring of Task at some point.